include/charmlite/core/ep.hpp

```diff
@@ -27,20 +27,121 @@ namespace cmk {
     template <auto Fn>
     struct entry_fn_impl_
     {
+    private:
+        template <typename... Args_>
+        static void func_invoker_impl_helper(void* self, Args_&&... args)
+        {
+            using template_t =
+                typename cmk::generate_marshall_msg<decltype(Fn)>::template_t;
+            (static_cast<template_t*>(self)->*Fn)(std::forward<Args_>(args)...);
+        }
+
+        template <typename Tuple, std::size_t... Indices>
+        static void func_invoker_impl(
+            void* self, Tuple&& t, std::index_sequence<Indices...>)
+        {
+            func_invoker_impl_helper(
+                self, std::get<Indices>(std::forward<Tuple>(t))...);
+        }
+
+        template <typename Tuple>
+        static void func_invoker(void* self, Tuple&& t)
+        {
+            func_invoker_impl(self, std::forward<Tuple>(t),
+                std::make_index_sequence<
+                    std::tuple_size<std::decay_t<Tuple>>::value>());
+        }
+
+        template <typename Serializer, typename T1, typename... Ts>
+        static void args_unfolder_impl_helper(
+            Serializer&& serializer, T1&& t1, Ts&&... ts)
+        {
+            serializer | (typename std::decay_t<T1>&) t1;
+
+            args_unfolder_impl_helper(
+                std::forward<Serializer>(serializer), std::forward<Ts>(ts)...);
+        }
+
+        template <typename Serializer, typename T1>
+        static void args_unfolder_impl_helper(Serializer&& serializer, T1&& t1)
+        {
+            serializer | (typename std::decay_t<T1>&) t1;
+        }
+
+        template <typename Serializer, typename Tuple, std::size_t... Indices>
+        static void args_unfolder_impl(
+            Serializer&& serializer, Tuple&& t, std::index_sequence<Indices...>)
+        {
+            args_unfolder_impl_helper(std::forward<Serializer>(serializer),
+                std::get<Indices>(std::forward<Tuple>(t))...);
+        }
+
+        template <typename Serializer, typename Tuple>
+        static void args_unfolder(Serializer&& serializer, Tuple&& ts)
```

Comment:
Can we move this into a header where other classes can use it?

Do we need this when we have `std::apply` in C++17? You can call a function with a tuple as args using it.

Reply:
I'm not sure if it can be used with a member function. The implementation of `std::apply` uses `template <class Fn, class Tuple>` so using it with member functions can lead to "call to non-static function" types of issues.
(I don't have much experience using `std::apply` so I may be wrong here)
A simple turn-around for it would be to use a lambda instead but I find this solution more "generalized".